src/Juvix/Backends/Michelson/Compilation/Lambda.hs

```diff
@@ -0,0 +1,174 @@
+  m Op
+termToInstrOuter term ty = do
+  maybeOp ← termToInstr term ty
+  case maybeOp of
+    Right op → pure op
+    Left (LamPartial ops captures args body _) → do
+      -- TODO: Actually compile the lambda to a closure.
+      -- We should never need to do this elsewhere.
+      -- ergo, if we do not return a lambda, we should never
+      -- compile to a lambda, except for the michelson built-ins
+      -- which take lambdas
+      -- todo: formalise this a bit
+      -- todo: deal with michelson builtins which take lambdas
+      -- they will have to use this function or something
+      undefined
+
+funcToLambda ∷
+  ∀ m.
+  ( HasState "stack" VStack.T m,
+    HasThrow "compilationError" CompilationError m,
+    HasWriter "compilationLog" [CompilationLog] m
+  ) ⇒
+  LamPartial →
+  M.Type →
+  m M.ExpandedOp
+funcToLambda (LamPartial ops captures args body lamTy) paramTy = do
+  -- ~~
+  -- Here we are dealing with a (possibly previously partially applied)
+  -- function which will not be fully applied, ergo we must compile it
+  -- to a Michelson lambda, which we do by copying the captures to the
+  -- top of the stack, packing them all in a tuple, compiling the body
+  -- of the lambda to a lambda function, then using the `APPLY` instruction
+  -- to partially apply the function & returning a lambda encapsulated
+  -- with this environment. Note that previous partial application of a
+  -- function to some arguments will result in those arguments being captured.
+  -- ~~
+  -- TODO ∷ what if all captures or arguments are constants? This will end horribly!
+  -- ~~
+  -- Step 1: Copy the captures to the top of the stack.
+  captureInsts ←
+    traverse
+      ( \x → do
+          currentStack ← get @"stack"
+          case VStack.lookup x currentStack of
+            Nothing →
+              failWith
+                ( "free variable in lambda"
+                    <> " doesn't exist"
+                )
+            Just (VStack.Value v) → do
+              let (Just type') = VStack.lookupType x currentStack
+              modify @"stack"
+                ( cons
+                    ( VStack.varE x (Just (VStack.ConstE v)),
+                      type'
+                    )
+                )
+              pure (M.SeqEx [])
+            Just (VStack.Position p) → do
+              let (Just type') = VStack.lookupType x currentStack
+              let inst = dupToFront (fromIntegral p)
+              modify @"stack" (cons (VStack.varE x Nothing, type'))
+              pure inst
+      )
+      captures
+  -- Step 2: Figure out what the stack will be in the body of the function.
+  current@(VStack.T currentStack _) ← get @"stack"
+  let realValuesGen xs =
+        length
+          $ filter (VStack.inT . fst)
+          $ take (length xs) currentStack
+
+      realValues = realValuesGen captures
+
+      numVarsInClosure = (length (captures <> args))
+
+  extraArgsWithTypes ← zip args . drop (length args) <$> typesFromPi lamTy
```

Comment:
Why are we dropping `(length args)`, that seems wrong now (maybe this was old logic)